crates/red_knot/src/hir.rs

```diff
@@ -0,0 +1,227 @@
+//! Key observations
+//!
+//! The HIR avoids allocations to large extends by:
+//! * Using an arena per node type
```

Comment:
The `Module` has an `IndexVec` for each node type which "acts" as an arena because we don't make a heap allocation for each node but instead make a single vector allocation and write each node into that vec.